types/jest-when/index.d.ts

```diff
@@ -10,21 +10,41 @@
 
 /// <reference types="jest" />
 
-export interface WhenMock<T = any, Y extends any[] = any>
-  extends jest.MockInstance<T, Y> {
-  calledWith(...matchers: Y): this;
-  expectCalledWith(...matchers: Y): this;
-  mockReturnValue(value: T): this;
-  mockReturnValueOnce(value: T): this;
-  mockResolvedValue(value: jest.ResolvedValue<T>): this;
-  mockResolvedValueOnce(value: jest.ResolvedValue<T>): this;
-  mockRejectedValue(value: jest.RejectedValue<T>): this;
-  mockRejectedValueOnce(value: jest.RejectedValue<T>): this;
-  mockImplementation(fn: (...args: Y) => T): this;
-  mockImplementationOnce(fn?: (...args: Y) => T): this;
+export type ArgumentOrMatcher<ArgTypes extends any[]> = {
+    [Index in keyof ArgTypes]: ArgTypes[Index] | WhenMock<boolean, [ArgTypes[Index]]>;
```

Comment:
Ideally, we wouldn't use `WhenMock` here but instead overload `when` to return another type (e.g. `WhenMatcher`) when it's called with a bare callable instead of a `jest.MockInstance`, and only accept those `WhenMatcher`s here. Reason being that [the argument callable for such a matcher cannot be a mock or spy](https://github.com/timkindberg/jest-when#supports-function-matchers) (see overloads [here](https://github.com/timkindberg/jest-when/blob/d4247dc4dddecd500dc4256490dee42f5e989d56/src/when.js#L164)). However, the typings currently allow us to do the following:
```javascript
const fn = jest.fn();
when(fn).calledWith(when(jest.fn())).mockReturnValue('foo');
```
which won't do the right thing.

Unfortunately, there's (AFAICT) no way to make that work with with #52924.

Reply:
I can’t tell from this comment—are changes to this PR necessary?

Reply:
I was just adding a note to my own PR 🙂 In summary: this works, but it's not as type-safe as it could be. However, making it type-safe would undo the changes of #52924 and re-introduce the inconvenience that that PR solved. Whether the PR needs changing depends on what's most important: type safety or usability.

Reply:
Got it. Sounds like we need package owners to weigh in. @aldentaylor, @immanuel192, @gstamac, @HipsterBrown